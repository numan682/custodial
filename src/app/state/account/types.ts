import { Transaction } from 'app/state/transaction/types'
import { ErrorPayload } from 'types/errors'
import { StringifiedBigInt } from 'types/StringifiedBigInt'

export interface BalanceDetails {
  available: StringifiedBigInt | null
  /** This is delayed in getAccount by 20 seconds on oasisscan and 5 seconds on oasismonitor. */
  debonding: StringifiedBigInt | null
  /** This is delayed in getAccount by 20 seconds on oasisscan and 5 seconds on oasismonitor. */
  delegations: StringifiedBigInt | null
  /** This is delayed in getAccount by 20 seconds on oasisscan and 5 seconds on oasismonitor. */
  total: StringifiedBigInt | null
}

export interface Allowance {
  address: string
  amount: StringifiedBigInt
}

export interface Account extends BalanceDetails {
  address: string
  allowances?: Allowance[]
}

/* --- STATE --- */
export interface AccountState extends Account {
  loading: boolean
  accountError?: ErrorPayload
  transactions: Transaction[]
  transactionsError?: ErrorPayload
}
