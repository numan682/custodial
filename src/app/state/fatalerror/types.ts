/* --- STATE --- */
export interface FatalErrorPayload {
  message: string
  stack?: string
  sagaStack?: string
}

export interface FatalErrorState {
  error?: FatalErrorPayload
}
