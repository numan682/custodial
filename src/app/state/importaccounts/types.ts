import { ErrorPayload } from 'types/errors'
import { WalletType } from '../wallet/types'
import { BalanceDetails } from '../account/types'

/* --- STATE --- */
export interface ImportAccountsListAccount {
  address: string
  balance?: BalanceDetails
  path: number[]
  pathDisplay: string
  privateKey?: string
  publicKey: string
  selected: boolean
  type: WalletType
}

export enum ImportAccountsStep {
  Idle = 'idle',
  OpeningUSB = 'opening_usb',
  LoadingAccounts = 'loading_accounts',
  LoadingBalances = 'loading_balances',
}

export interface ImportAccountsState {
  accounts: ImportAccountsListAccount[]
  error?: ErrorPayload
  showAccountsSelectionModal: boolean
  accountsSelectionPageNumber: number
  step: ImportAccountsStep
}
