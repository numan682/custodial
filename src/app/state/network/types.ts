/* --- STATE --- */

export type NetworkType = 'local' | 'testnet' | 'mainnet'

export interface NetworkState {
  selectedNetwork: NetworkType

  /** Ticker symbol (ROSE, TEST, etc.) */
  ticker: string

  /** chainContext / Genesis Hash */
  chainContext: string

  /** Current epoch */
  epoch: number

  /** Minimum staking amount */
  minimumStakingAmount: number
}
