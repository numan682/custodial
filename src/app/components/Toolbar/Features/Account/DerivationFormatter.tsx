import { WalletType } from 'app/state/wallet/types'
import { Box } from 'grommet/es6/components/Box'
import { Text } from 'grommet/es6/components/Text'
import { useTranslation } from 'react-i18next'

export interface DerivationFormatterProps {
  type: WalletType
  pathDisplay: string | undefined
}

export const DerivationFormatter = (props: DerivationFormatterProps) => {
  const { t } = useTranslation()
  const walletTypes: { [type in WalletType]: string } = {
    [WalletType.Ledger]: t('toolbar.wallets.type.ledger', 'Ledger'),
    [WalletType.Mnemonic]: t('toolbar.wallets.type.mnemonic', 'Mnemonic'),
    [WalletType.PrivateKey]: t('toolbar.wallets.type.privateKey', 'Private key'),
  }
  return (
    <Box align="start" direction="row">
      {walletTypes[props.type]} {props.pathDisplay && <Text size="80%">({props.pathDisplay})</Text>}
    </Box>
  )
}
