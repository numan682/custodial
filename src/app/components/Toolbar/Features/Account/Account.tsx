import { AmountFormatter } from 'app/components/AmountFormatter'
import { PrettyAddress } from 'app/components/PrettyAddress'
import { ShortAddress } from 'app/components/ShortAddress'
import { Box } from 'grommet/es6/components/Box'
import { CheckBox } from 'grommet/es6/components/CheckBox'
import { Spinner } from 'grommet/es6/components/Spinner'
import { Text } from 'grommet/es6/components/Text'
import { ResponsiveContext } from 'grommet/es6/contexts/ResponsiveContext'
import { memo, useContext } from 'react'
import { useTranslation } from 'react-i18next'
import { BalanceDetails } from '../../../../state/account/types'
import { Button } from 'grommet/es6/components/Button'
import { DerivationFormatter, DerivationFormatterProps } from './DerivationFormatter'

interface AccountProps {
  address: string
  balance: BalanceDetails | undefined
  onClick: (address: string) => void
  path?: number[]
  isActive: boolean
  displayCheckbox?: boolean
  displayAccountNumber?: boolean
  displayDerivation?: DerivationFormatterProps
  displayManageButton?: {
    onClickManage: (address: string) => void
  }
}

export const Account = memo((props: AccountProps) => {
  const { t } = useTranslation()
  const size = useContext(ResponsiveContext)

  const address =
    size === 'small' ? <ShortAddress address={props.address} /> : <PrettyAddress address={props.address} />

  const accountNumber = props.path ? props.path.at(-1) : '?'

  return (
    <Box
      data-testid="account-choice"
      round="5px"
      background={props.isActive ? 'neutral-2' : undefined}
      border={{ color: props.isActive ? 'neutral-2' : 'brand' }}
      pad="small"
      flex="grow"
      fill="horizontal"
      role="checkbox"
      aria-checked={props.isActive}
      onClick={() => {
        props.onClick(props.address)
      }}
      hoverIndicator={{ background: 'brand' }}
      direction="row"
    >
      {props.displayCheckbox && (
        <Box alignSelf="center" pad={{ left: 'small', right: 'medium' }}>
          <CheckBox checked={props.isActive} hidden />
        </Box>
      )}
      {props.displayAccountNumber && (
        <Box alignSelf="center" pad={{ left: 'small', right: 'small' }} style={{ minWidth: '2.5em' }}>
          <Text weight="bold" style={{ width: '2em' }}>
            {accountNumber}
          </Text>
        </Box>
      )}
      <Box flex="grow" gap={size === 'small' ? undefined : 'xsmall'}>
        <Box>
          <Text weight="bold">{address}</Text>
        </Box>
        <Box direction="row-responsive">
          <Box flex="grow">
            {props.displayDerivation && <DerivationFormatter {...props.displayDerivation} />}
            {props.displayManageButton && (
              <Box direction="row">
                <Button
                  label={t('toolbar.settings.manageAccount', 'Manage')}
                  onClick={e => {
                    props.displayManageButton?.onClickManage(props.address)
                    e.stopPropagation()
                  }}
                />
              </Box>
            )}
          </Box>
          <Box height={'24px'}>
            {props.balance ? <AmountFormatter amount={props.balance.total} /> : <Spinner />}
          </Box>
        </Box>
      </Box>
    </Box>
  )
})
