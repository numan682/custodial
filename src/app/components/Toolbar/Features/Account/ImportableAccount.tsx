import { Account } from './Account'
import { ImportAccountsListAccount } from '../../../../state/importaccounts/types'

export const ImportableAccount = ({
  account,
  onClick,
}: {
  account: ImportAccountsListAccount
  onClick: (address: string) => void
}) => {
  return (
    <Account
      address={account.address}
      balance={account.balance}
      onClick={onClick}
      isActive={account.selected}
      displayCheckbox={true}
      displayAccountNumber={true}
      displayDerivation={{
        type: account.type,
        pathDisplay: account.pathDisplay,
      }}
      path={account.path}
    />
  )
}
