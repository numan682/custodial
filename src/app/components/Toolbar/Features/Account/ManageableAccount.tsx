import { Box } from 'grommet/es6/components/Box'
import { Button } from 'grommet/es6/components/Button'
import { Tab } from 'grommet/es6/components/Tab'
import { Text } from 'grommet/es6/components/Text'
import { ResponsiveContext } from 'grommet/es6/contexts/ResponsiveContext'
import { useContext, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Account } from './Account'
import { Wallet } from '../../../../state/wallet/types'
import { ResponsiveLayer } from '../../../ResponsiveLayer'
import { Tabs } from 'grommet/es6/components/Tabs'
import { DerivationFormatter } from './DerivationFormatter'
import { uintToBase64, hex2uint } from '../../../../lib/helpers'
import { AddressBox } from '../../../AddressBox'

export const ManageableAccount = ({
  wallet,
  isActive,
  onClick,
}: {
  wallet: Wallet
  isActive: boolean
  onClick: (address: string) => void
}) => {
  const { t } = useTranslation()
  const [layerVisibility, setLayerVisibility] = useState(false)
  const isMobile = useContext(ResponsiveContext) === 'small'
  return (
    <>
      <Account
        address={wallet.address}
        balance={wallet.balance}
        onClick={onClick}
        isActive={isActive}
        path={wallet.path}
        displayManageButton={{
          onClickManage: () => setLayerVisibility(true),
        }}
      />
      {layerVisibility && (
        <ResponsiveLayer
          onClickOutside={() => setLayerVisibility(false)}
          onEsc={() => setLayerVisibility(false)}
          animation="none"
          background="background-front"
          modal
          position="top"
          margin={isMobile ? 'none' : 'xlarge'}
        >
          <Box margin="medium" width={isMobile ? 'auto' : '700px'}>
            <Tabs alignControls="start">
              <Tab title={t('toolbar.settings.myAccountsTab', 'My Accounts')}>
                <Box margin={{ vertical: 'medium' }}>
                  <AddressBox address={wallet.address} border />
                  <Text size="small" margin={'small'}>
                    <DerivationFormatter pathDisplay={wallet.pathDisplay} type={wallet.type} />
                  </Text>
                </Box>
                <Button
                  label={t('toolbar.settings.exportPrivateKey', 'Export Private Key')}
                  disabled={!wallet.privateKey}
                  onClick={() => {
                    prompt(
                      t('toolbar.settings.exportPrivateKey', 'Export Private Key'),
                      uintToBase64(hex2uint(wallet.privateKey!)),
                    )
                  }}
                />
                <Box direction="row" justify="between" pad={{ top: 'large' }}>
                  <Button
                    secondary
                    label={t('toolbar.settings.cancel', 'Cancel')}
                    onClick={() => setLayerVisibility(false)}
                  />
                </Box>
              </Tab>
            </Tabs>
          </Box>
        </ResponsiveLayer>
      )}
    </>
  )
}
