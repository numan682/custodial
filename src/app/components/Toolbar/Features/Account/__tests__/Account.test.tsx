import { render } from '@testing-library/react'
import { WalletType } from 'app/state/wallet/types'
import * as React from 'react'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'
import { ThemeProvider } from 'styles/theme/ThemeProvider'

import { Account } from '../Account'

const renderComponent = (store: any) =>
  render(
    <Provider store={store}>
      <ThemeProvider>
        <Account
          address="oasis1qq3xrq0urs8qcffhvmhfhz4p0mu7ewc8rscnlwxe"
          balance={{ available: '200', debonding: '0', delegations: '800', total: '1000' }}
          onClick={() => {}}
          isActive={false}
          displayCheckbox={true}
          displayAccountNumber={true}
          path={[44, 474, 0, 0, 0]}
          displayDerivation={{
            type: WalletType.Mnemonic,
            pathDisplay: "m/44'/474'/0'/0'/0'",
          }}
        />
      </ThemeProvider>
    </Provider>,
  )

describe('<Account  />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore({})
  })

  it('should match snapshot', () => {
    const component = renderComponent(store)
    expect(component.baseElement).toMatchSnapshot()
  })
})
