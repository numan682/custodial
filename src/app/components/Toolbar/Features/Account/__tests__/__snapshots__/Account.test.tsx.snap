// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account  /> should match snapshot 1`] = `
.c0 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #16213E;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c9 {
  font-size: 18px;
  line-height: 24px;
  font-weight: bold;
}

.c16 {
  font-size: 18px;
  line-height: 24px;
}

.c14 {
  font-size: 80%;
  line-height: normal;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border: solid 1px #0092f6;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  width: 100%;
  padding: 12px;
  border-radius: 5px;
  cursor: pointer;
}

.c1:hover {
  background-color: #0092f6;
  color: #FFFFFF;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-left: 12px;
  padding-right: 24px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: solid 2px rgba(0,0,0,0.15);
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 24px;
  width: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-left: 12px;
  padding-right: 12px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 24px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.c3:hover input:not([disabled]) + div,.c3:hover input:not([disabled]) + span {
  border-color: #000000;
}

.c6 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

.c6:checked + span > span {
  left: calc( 48px - 24px );
  background: #0092f6;
}

.c5 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

@media only screen and (max-width:768px) {
  .c1 {
    border: solid 1px #0092f6;
  }
}

@media only screen and (max-width:768px) {
  .c1 {
    padding: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    padding-left: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    padding-right: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c7 {
    border: solid 2px rgba(0,0,0,0.15);
  }
}

@media only screen and (max-width:768px) {
  .c8 {
    padding-left: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c8 {
    padding-right: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c12 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-flex-basis: auto;
    -ms-flex-preferred-size: auto;
    flex-basis: auto;
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    -webkit-align-items: stretch;
    -webkit-box-align: stretch;
    -ms-flex-align: stretch;
    align-items: stretch;
  }
}

@media only screen and (max-width:768px) {

}

<body>
  <div>
    <div
      class="c0"
      style="min-height: 100vh;"
    >
      <div
        aria-checked="false"
        class="c1"
        data-testid="account-choice"
        role="checkbox"
        tabindex="0"
      >
        <div
          class="c2"
        >
          <label
            class="c3"
          >
            <div
              class="c4 c5"
            >
              <input
                class="c6"
                hidden=""
                type="checkbox"
              />
              <div
                class="c7 "
              />
            </div>
          </label>
        </div>
        <div
          class="c8"
          style="min-width: 2.5em;"
        >
          <span
            class="c9"
            style="width: 2em;"
          >
            0
          </span>
        </div>
        <div
          class="c10"
        >
          <div
            class="c11"
          >
            <span
              class="c9"
            >
              <span
                class="notranslate"
                translate="no"
              >
                oasis1qq3x...rscnlwxe
              </span>
            </span>
          </div>
          <div
            class="c12"
          >
            <div
              class="c10"
            >
              <div
                class="c13"
              >
                toolbar.wallets.type.mnemonic
                 
                <span
                  class="c14"
                >
                  (
                  m/44'/474'/0'/0'/0'
                  )
                </span>
              </div>
            </div>
            <div
              class="c15"
            >
              <span>
                <span>
                  0.000001
                </span>
                <span
                  class="c16"
                >
                  <span
                    class="notranslate"
                    translate="no"
                  >
                     
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
